### PR TITLE
added structured 8_keras_dnn_babyweight

### PR DESCRIPTION
**Pull request by ryangillard**

### CODE REVIEW

File: courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb [hunk 1]
```diff
@@ -0,0 +1,574 @@
+{
```
Comment (munnm) [inline courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: <p>The "In this lab,..." paragraph feels very run-on to me. I'd suggest breaking it up into multiple sentences. Something along the lines of</p><p> </p><p>"We'll start by defining the CSV column names, label column and column defaults for our data inputs. Then, we'll construct a tf.data Dataset of features and the label from the CSV files and create inputs layers for the raw features. Next, we'll set up feature columns for the model inputs and build a deep neural network in Keras. We'll create a custom evaluation metric and build our DNN model and finally train and evaluate our model."</p><p> </p><p>If re-written, adjust in solution notebook to match.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/672/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (munnm) [inline courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: <p>why not combine these three commands to a single</p><p>%%bash</p><p>head -5 *.csv</p><p> </p><p>I think it still shows the result fairly concisely. Just personal preference.</p><p> </p><p>If changed, adjust in solution notebook as well.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/672/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (munnm) [inline courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: <p>the description of DEFAULTS took me a second.</p><p> </p><p>Should it be "DEFAULTS is a list of length equal to the length CSV_COLUMNS". Each element of DEFAULTS is a list itself with the default value for the respective CSV column.</p><p> </p><p>If re-written, change in solution notebook to match.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/672/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (munnm) [inline courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: <p>Nit. Appears that other Lab Tasks use ### header instead of ## header. Replace with ### here for consistency. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/672/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

File: courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb [hunk 2]
```diff
@@ -0,0 +1,784 @@
+{
```
Comment (munnm) [inline courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: <p>I know in other labs we've used imports like</p><p>from tensorflow.keras.layers import (Input, Dense, etc</p><p>to simplify later code. For example, here (and elsewhere) would allow each layer to fit on one code line.</p><p> </p><p>(similarly when importing feature columns; e.g. from tensorflow import feature_column as fc)</p><p> </p><p>I don't think it's strictly necessary here since it's fairly simple enough. And personally it can be nice to see the full 'tf.keras.layers.Dense..." or "tf.feature_column.numeric_column..." especially in a learning scenario.</p><p> </p><p>Up to you....</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/672/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (munnm) [inline courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: <p>nit. I prefer **2 over tf.square.</p><p> </p><p>Curious, why use tf.square instead?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/672/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (munnm) [inline courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: <p>This wasn't working for me. But think this is a CAIP problem</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/672/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (munnm) [inline courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: <p>same as above. i'd suggest rewording. But just personal preference.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/672/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

**Comment by munnm** [comment]
<p>hey Ryan, nice lab/solution. Most of my comments are expository or personal preference. Looks good overall. Happy to discuss more. </p>

File: courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb (hunk 1 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: <p>Thanks, ya I was trying to match my other notebooks of having the learning objectives in that sentence. But this notebook has a ton of them so it did get rather long. Changed to close to what you suggested!</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: <p>Ya, that's fair. Less cells to run and exactly the same output. I just wanted to anyways to be a simple, "Look everything is there, yay!" thing which this will do too. Changed.</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: <p>I had 1-7 as ### because they were underneath ## Create Keras model heading. Train and evaluate was outside of that "heading scope" therefore I went back to ## heading type for it. So I added ## Run and evaluate model  above it so that I could downgrade task 8 to ###. I think it looks better now, at least my systematic :)</p>

File: courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb (hunk 2 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: <p>Thanks, done.</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: <p>Whenever I've taught the couple examples that have it the short imported alias, at least my students didn't like it. Having fc.X wasn't reminding them of where it was from or what fc stood for. Also, I personally prefer the long way myself to be abundantly clear where it comes from and saves me from having to scroll back to the top to remember what all of the acronyms mean haha.</p>

File: courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb (hunk 1 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: <p>Ya, I changed it, it was a little convoluted sounding, thanks!</p>

File: courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb (hunk 2 above)
Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: <p>Ya, I was getting a 500 error too. I tried it both in 1.15 and 2.0 initialized CAIP notebooks. I think it must be some CAIP bug.</p>

Comment (ryangillard) [reply courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: <p>Done. I don't know, maybe leftover from C with math.pow() haha</p>

**Comment by ryangillard** [comment]
<p>Thanks for the review Mike! I responded to all of your comments.</p>

File: courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb (hunk 2 above)
Comment (munnm) [reply courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: yeah. I understand. I think it's a bit of a trade-off between concise code and easy of reading for instruction purposes. In this situation it's already fairly contained so happy to leave as is.

File: courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb (hunk 1 above)
Comment (munnm) [reply courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: ok.

Comment (munnm) [reply courses/machine_learning/deepdive2/structured/labs/4b_keras_dnn_babyweight.ipynb]: ok. yeah fair enough. thanks.

File: courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb (hunk 2 above)
Comment (munnm) [reply courses/machine_learning/deepdive2/structured/solutions/4b_keras_dnn_babyweight.ipynb]: ok. I guess we should leave as is now and wait for CAIP to catch up. We definitely want Tensorboard used here.
